src/contracts/core/DelegationManager.sol

```diff
@@ -528,132 +560,174 @@ contract DelegationManager is
+    /// TODO: natspec
+    function _completeReceiveAsTokens(
+        Withdrawal calldata withdrawal,
+        IERC20[] calldata tokens,
+        uint256[] memory descaledShares
+    ) internal {
+        // Finalize action by converting scaled shares to tokens for each strategy, or
+        // by re-awarding shares in each strategy.
+        for (uint256 i = 0; i < withdrawal.strategies.length;) {
+            require(
+                withdrawal.startTimestamp + strategyWithdrawalDelays[withdrawal.strategies[i]] <= block.timestamp,
+                "DelegationManager._completeReceiveAsTokens: withdrawalDelay period has not yet passed for this strategy"
+            );
 
-                _withdrawSharesAsTokens({
-                    staker: withdrawal.staker,
-                    withdrawer: msg.sender,
-                    strategy: withdrawal.strategies[i],
-                    shares: withdrawal.shares[i],
-                    token: tokens[i]
-                });
-                unchecked {
-                    ++i;
-                }
+            _withdrawSharesAsTokens({
+                staker: withdrawal.staker,
+                withdrawer: msg.sender,
+                strategy: withdrawal.strategies[i],
+                shares: descaledShares[i],
+                token: tokens[i]
+            });
+            unchecked {
+                ++i;
             }
-            // Award shares back in StrategyManager/EigenPodManager. If withdrawer is delegated, increase the shares delegated to the operator
-        } else {
-            // Award shares back in StrategyManager/EigenPodManager.
-            // If withdrawer is delegated, increase the shares delegated to the operator.
-            address currentOperator = delegatedTo[msg.sender];
-            for (uint256 i = 0; i < withdrawal.strategies.length;) {
-                require(
-                    withdrawal.startBlock + strategyWithdrawalDelayBlocks[withdrawal.strategies[i]] <= block.number,
-                    "DelegationManager._completeQueuedWithdrawal: withdrawalDelayBlocks period has not yet passed for this strategy"
-                );
+        }
+    }
+
+    /// TODO: natspec
+    function _completeReceiveAsShares(
+        Withdrawal calldata withdrawal,
+        IERC20[] calldata tokens,
+        uint256[] memory descaledShares
+    ) internal {
+        // read delegated operator for scaling and adding shares back if needed
+        address currentOperator = delegatedTo[msg.sender];
+        // We scale shares again to the new totalMagnitude of the currentOperator.
+        uint256[] memory scaledShares = ShareScalingLib.scaleShares(
+            avsDirectory,
+            currentOperator,
+            withdrawal.strategies,
+            descaledShares
+        );
 
+        for (uint256 i = 0; i < withdrawal.strategies.length;) {
+            require(
+                withdrawal.startTimestamp + strategyWithdrawalDelays[withdrawal.strategies[i]] <= block.number,
```

Comment:
This doesn't seem to work for uncompleted withdrawals from before the upgrade?

Reply:
gp, added to pr description